### PR TITLE
Feat autonoise

### PR DESCRIPTION
**Pull request by jtwhite79**
added `helpers.py::autocorrelated_draws()` to help generate better noise reals.  Also a few small patches in `PstFrom` when using direct parameters, weaving thru `initial_value`, better defaults for pilot point specific args for "none" transform, and also detecting when (and when not) to reuse pilot point factors

### CODE REVIEW

File: pyemu/utils/helpers.py [hunk 1]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+def autocorrelated_draw(pst,struct_dict,time_distance_col="distance",num_reals=100,verbose=True,
+                        enforce_bounds=False, draw_ineq=False):
+    """construct an autocorrelated observation noise ensemble from covariance matrices
+        implied by geostatistical structure(s).
+
+        Args:
+            pst (`pyemu.Pst`): a control file (or the name of control file).  The
+                information in the `* observation data` dataframe is used extensively,
+                including weight, standard_deviation (if present), upper_bound/lower_bound (if present).
+            time_distance_col (str): the column in `* observation_data` the represnts the distance in time
```
Comment (briochh) [inline pyemu/utils/helpers.py:102]:
```suggestion
            time_distance_col (str): the column in `* observation_data` that represents the distance in time
```

File: pyemu/utils/helpers.py [hunk 2]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+    if enforce_bounds:
+        if verbose:
+            print("-->enforcing bounds")
+        ub_dict = {o:1e300 for o in full_oe.columns}
+        if "upper_bound" in pst.observation_data.columns:
+            ub_dict.update(pst.observation_data.upper_bound.fillna(1.0e300).to_dict())
+
+        lb_dict = {o:-1e300 for o in full_oe.columns}
+        if "lower_bound" in pst.observation_data.columns:
+            lb_dict.update(pst.observation_data.lower_bound.fillna(-1.0e200).to_dict())
```
Comment (briochh) [inline pyemu/utils/helpers.py:190]:
```suggestion
            lb_dict.update(pst.observation_data.lower_bound.fillna(-1.0e300).to_dict())
```

File: pyemu/utils/helpers.py [hunk 3]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+            full_oe.loc[:,n] = v
+        obs = pst.observation_data
+        gt_tags = pst.get_constraint_tags("gt")
+        gt_onames = [oname for oname, ogrp in zip(obs.obsnme, obs.obgnme) if
+                     True in [True if str(ogrp).startswith(tag) else False for tag in gt_tags]]
+        if verbose:
+            print("--> greater than ineq obs:", gt_onames)
+        gt_dict = obs.loc[gt_onames, "obsval"].to_dict()
+        for n, v in gt_dict.items():
+            full_oe.loc[:, n] = v
```
Comment (briochh) [inline pyemu/utils/helpers.py:219]: dito above direct assignment

File: pyemu/utils/helpers.py [hunk 4]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+        if verbose:
+            print("--> less than ineq obs:",lt_onames)
+        lt_dict = obs.loc[lt_onames,"obsval"].to_dict()
+        for n,v in lt_dict.items():
+            full_oe.loc[:,n] = v
+        obs = pst.observation_data
+        gt_tags = pst.get_constraint_tags("gt")
+        gt_onames = [oname for oname, ogrp in zip(obs.obsnme, obs.obgnme) if
+                     True in [True if str(ogrp).startswith(tag) else False for tag in gt_tags]]
+        if verbose:
```
Comment (briochh) [inline pyemu/utils/helpers.py:215]: ditto above, could use `pst.greater_than_obs_constraints()` option

File: pyemu/utils/helpers.py [hunk 5]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+            lb_dict.update(pst.observation_data.lower_bound.fillna(-1.0e200).to_dict())
+        allvals = full_oe.values
+        for i,name in enumerate(full_oe.columns):
+            #print("before:",name,ub_dict[name],full_oe.loc[:,name].max(),lb_dict[name],full_oe.loc[:,name].min())
+            #vals = full_oe.loc[:,name].values
+            vals = allvals[:,i]
+            vals[vals>ub_dict[name]] = ub_dict[name]
+            vals[vals < lb_dict[name]] = lb_dict[name]
+            #full_oe.loc[:,name] = vals#oe.loc[:,name].apply(lambda x: min(x,ub_dict[name])).apply(lambda x: max(x,lb_dict[name]))
+            allvals[:,i] = vals
```
Comment (briochh) [inline pyemu/utils/helpers.py:199]: do these ever get passed back to full_oe?

**Review by briochh — CHANGES_REQUESTED** [review]
a few places changes to ensemble values may not be getting passed back to returned ensemble (e.g. when applying bounds).

might be possible to reduce looping -- could imagine that this might get slow where nobs is big

File: pyemu/mat/mat_handler.py [hunk 6]
```diff
@@ -3119,12 +3119,17 @@ def from_observation_data(cls, pst):
         x = np.zeros((nobs, 1))
         onames = []
         ocount = 0
+        std_dict = {}
+        if "standard_deviation" in pst.observation_data.columns:
+            std_dict = pst.observation_data.standard_deviation.to_dict()
+            std_dict = {o:s**2 for o,s in std_dict.items() if pd.notna(s)}
```
[the remark this answers is not in the document]
Comment (jtwhite79) [reply pyemu/mat/mat_handler.py:3125]: this is what we want - the `standard_deviation` col is what can be passed on the dataframe but we need var for the cov matrix

File: pyemu/utils/helpers.py [hunk 7]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+        """
+
+    #check that the required time metadata is appropriate
+    passed_names = []
+    nz_names = pst.nnz_obs_names
+    [passed_names.extend(obs) for gs,obs in struct_dict.items()]
+    missing = list(set(passed_names) - set(nz_names))
+    if len(missing) > 0:
+        raise Exception("the following obs in struct_dict were not found in the nz obs names"+str(missing))
+    time_cols = ["time","datetime","distance"]
```
[the remark this answers is not in the document]
Comment (jtwhite79) [reply pyemu/utils/helpers.py:149]: yeah this is a left over from earlier iterations.  will remove

File: pyemu/utils/helpers.py [hunk 8]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+        ub_dict = {o:1e300 for o in full_oe.columns}
+        if "upper_bound" in pst.observation_data.columns:
+            ub_dict.update(pst.observation_data.upper_bound.fillna(1.0e300).to_dict())
+
+        lb_dict = {o:-1e300 for o in full_oe.columns}
+        if "lower_bound" in pst.observation_data.columns:
+            lb_dict.update(pst.observation_data.lower_bound.fillna(-1.0e200).to_dict())
+        allvals = full_oe.values
+        for i,name in enumerate(full_oe.columns):
+            #print("before:",name,ub_dict[name],full_oe.loc[:,name].max(),lb_dict[name],full_oe.loc[:,name].min())
```
[the remark this answers is not in the document]
Comment (jtwhite79) [reply pyemu/utils/helpers.py:193]: seemed to be but could use some testing on larger nobs

File: pyemu/utils/helpers.py [hunk 9]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+            vals[vals < lb_dict[name]] = lb_dict[name]
+            #full_oe.loc[:,name] = vals#oe.loc[:,name].apply(lambda x: min(x,ub_dict[name])).apply(lambda x: max(x,lb_dict[name]))
+            allvals[:,i] = vals
+            #print("...after:", name, ub_dict[name],full_oe.loc[:, name].max(),  lb_dict[name], full_oe.loc[:, name].min(), )
+
+    if not draw_ineq:
+        obs = pst.observation_data
+        lt_tags = pst.get_constraint_tags("lt")
+        lt_onames = [oname for oname,ogrp in zip(obs.obsnme,obs.obgnme) if True in [True if str(ogrp).startswith(tag) else False for tag in lt_tags]  ]
+        if verbose:
```
[the remark this answers is not in the document]
Comment (jtwhite79) [reply pyemu/utils/helpers.py:206]: I forgot about that method - will revise

File: pyemu/utils/helpers.py [hunk 10]
```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
...
+            print("--> less than ineq obs:",lt_onames)
+        lt_dict = obs.loc[lt_onames,"obsval"].to_dict()
+        for n,v in lt_dict.items():
+            full_oe.loc[:,n] = v
```
[the remark this answers is not in the document]
Comment (jtwhite79) [reply pyemu/utils/helpers.py:210]: It seemed to be fast enough for my needs but can always use some faste implementations!

File: pyemu/utils/helpers.py (hunk 3 above)
Comment (jtwhite79) [reply pyemu/utils/helpers.py:219]: will revise

File: pyemu/utils/helpers.py (hunk 4 above)
Comment (jtwhite79) [reply pyemu/utils/helpers.py:215]: will revise

File: pyemu/utils/helpers.py (hunk 5 above)
Comment (jtwhite79) [reply pyemu/utils/helpers.py:199]: I checked this and the ref to `full_oe.values` remained intact - the testing covers this

**Review by briochh — APPROVED** [review]
Sweet

**Pull request merged**